### ISSUE TITLE
Unsuccessful attack in scenario of UCF101 baseline pretrained

### ISSUE THREAD

**Issue by mzweilin** [twosixlabs/armory-example#35]
Hi there,

I have run the official UCF101 scenario example on GPU, but the result doesn't look right: the accuracy is the same for legitimate and adversarial examples:

```json
    "results": {
        "adversarial_mean_categorical_accuracy": 0.9275707110758658,
        "adversarial_mean_top_5_categorical_accuracy": 0.9875759978852763,
        "benign_mean_categorical_accuracy": 0.9275707110758658,
        "benign_mean_top_5_categorical_accuracy": 0.9875759978852763,
        "perturbation_mean_linf": 51.0
    }
```

Steps to reproduce the issue:

```diff
diff --git a/official_scenario_configs/ucf101_baseline_pretrained.json b/official_scenario_configs/ucf101_baseline_pretrained.json
index 9bb748f..17495f9 100644
--- a/official_scenario_configs/ucf101_baseline_pretrained.json
+++ b/official_scenario_configs/ucf101_baseline_pretrained.json
@@ -47,6 +47,6 @@
         "docker_image": "twosixarmory/pytorch:0.6.0",
         "external_github_repo": "yusong-tan/MARS",
         "gpus": "all",
-        "use_gpu": false
+        "use_gpu": true
     }
 }
```

```shell
$ armory run official_scenario_configs/ucf101_baseline_pretrained.json
```

The full json output is here:
```json
{
    "armory_version": "0.6.0",
    "config": {
        "_description": "UCF101 video classification from pretrained, contributed by MITRE Corporation",
        "adhoc": null,
        "attack": {
            "budget": null,
            "knowledge": "white",
            "kwargs": {
                "eps": 51.0
            },
            "module": "art.attacks",
            "name": "FastGradientMethod"
        },
        "dataset": {
            "batch_size": 16,
            "module": "armory.data.datasets",
            "name": "ucf101"
        },
        "defense": null,
        "metric": {
            "means": true,
            "perturbation": "linf",
            "record_metric_per_sample": false,
            "task": [
                "categorical_accuracy",
                "top_5_categorical_accuracy"
            ]
        },
        "model": {
            "fit": false,
            "fit_kwargs": {
                "nb_epochs": 10
            },
            "model_kwargs": {
                "model_status": "ucf101_trained"
            },
            "module": "armory.baseline_models.pytorch.ucf101_mars",
            "name": "get_art_model",
            "weights_file": "mars_ucf101_v1.pth",
            "wrapper_kwargs": {}
        },
        "scenario": {
            "kwargs": {},
            "module": "armory.scenarios.video_ucf101_scenario",
            "name": "Ucf101"
        },
        "sysconfig": {
            "docker_image": "twosixarmory/pytorch:0.6.0",
            "external_github_repo": "yusong-tan/MARS",
            "gpus": "all",
            "use_gpu": true
        }
    },
    "results": {
        "adversarial_mean_categorical_accuracy": 0.9275707110758658,
        "adversarial_mean_top_5_categorical_accuracy": 0.9875759978852763,
        "benign_mean_categorical_accuracy": 0.9275707110758658,
        "benign_mean_top_5_categorical_accuracy": 0.9875759978852763,
        "perturbation_mean_linf": 51.0
    }
}

```

**Comment by seanpmorgan** [comment]
Hi @mzweilin thanks for reporting! So I beleive this is the same issue as:
https://github.com/twosixlabs/armory/issues/396

The problem is that the evasion attack builds adversarial examples for single frames. It's not targeting the concatenation of the frames as it should (the true task). We just found this recently and were discussing possible options.

@yusong-tan @davidslater who might have thoughts on how we should proceed.

**Comment by davidslater** [comment]
The fix is relatively straightforward - it basically involves changing the network by adding a mean layer after the softmax output - which will then make the gradients correct for the attack. I'm trying to find a more general way to work with models that require modification after training.

**Comment by yusong-tan** [comment]
The issue raised by @mzweilin is due to a typo in the scenario where the clean examples, instead of adversarial examples, were passed to the classifier.predict().

The location of the typo is: https://github.com/twosixlabs/armory/blob/d6128f89e274f2e42b8a3a16379de984dfa9260e/armory/scenarios/video_ucf101_scenario.py#L124

With the change, the results are:

```json
"results": {
        "adversarial_mean_categorical_accuracy": 0.015596087761036215,
        "adversarial_mean_top_5_categorical_accuracy": 0.06502775574940524,
        "benign_mean_categorical_accuracy": 0.9275707110758658,
        "benign_mean_top_5_categorical_accuracy": 0.9875759978852763,
        "perturbation_mean_linf": 51.0
    }
```

**Comment by yusong-tan** [comment]
@davidslater Adding a mean layer could work.  The caveats to that are: it should only be done for inference and the inputs to the model during inference should come from the same source (i.e., video, audio, sentence, etc.).  The latter will need to be enforced somehow so people don't inadvertently batch inputs from multiple sources, which will then be erroneously averaged.

**Comment by davidslater** [comment]
@mzweilin I've added Mike's quickfix on the bug into a PR - that will be availabe in our next patch update.

@yusong-tan We're looking into a generic way to handle differences in training and inference time, of which the MARS model is one example. I think that the protecting against overlapping sources should be straightforward. The more generic part needs further thought, though.
